Introduce i.MX RT FlexSPI HyperRAM driver

pdgendt opened the pull request:
Add the FlexSPI HyperBUS driver to support HyperRAM external devices.

Technically not a flash driver so it currently doesn't have the flash API implemented. Can be added, if required.

Tested with [Cypress S27KS0641](https://www.cypress.com/file/183506/download) and following `dts` config:
```
&flexspi {
 	status = "okay";
	reg = <0x402a8000 0x4000>, <0x60000000 DT_SIZE_M(8)>;
	combination-mode;
	sck-differential-clock;
	rx-clock-source = <3>; // External input from DQS pad

	s27ks0641: s27ks0641@0 {
		status = "okay";
		compatible = "nxp,imx-flexspi-hyperram";
		reg = <0>;
		size = <67108864>;
		label = "S27KS0641";
		spi-max-frequency = <332000000>;

		cs-interval-unit = <1>;
		cs-interval = <2>;
		cs-setup-time = <4>;
		cs-hold-time = <0>;
		data-valid-time = <1>;
		column-space = <3>;
		word-addressable;
		ahb-write-wait-unit = <2>;
		ahb-write-wait-interval = <0>;
	};
};
```

pdgendt commented:
[This application note](https://www.nxp.com/docs/en/nxp/application-notes/AN12239.pdf) specifies how to improve read/write performance with MPU settings and `ahb` buffer/prefetch features.
But I don't know how to do the MPU setting properly and enabling `ahb-prefetch` on my custom board resulted in bus faults so there is probably something missing here to fully utilize the HyperRAM.

carlescufi commented:
@sylvioalves copying you here since you mentioned the same issue to me recently

carlescufi commented:
@erwango does STM32 have an equivalent? i.e. RAM connected over SPI, but then directly addressable after.

erwango commented:
> @erwango does STM32 have an equivalent? i.e. RAM connected over SPI, but then directly addressable after.

Yes, we also have OCTOSPI IPs supporting HyperRAM memories in memory mapped mode.

carlescufi commented:
API meeting 2021.03.30:

- Move `flash_mcux_flexspi.c` and `flash_mcux_flexspi_hyperram.c` to `drivers/memc`
- Leave `flash_mcux_flexspi_nor.c` in `drivers/flash`

pdgendt committed:
mcux flexspi: move bus driver to drivers/memc

Initially the flexspi device only supported a flash driver for
external NOR flash. As the controller supports HyperBus devices,
which can be either volatile or non-volatile, the driver iss moved
to drivers/memc.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
drivers: memc: Introduce i.MX RT FlexSPI HyperRAM driver

Add the FlexSPI HyperBUS driver to support HyperRAM external
devices.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

MaureenHelm commented on drivers/memc/Kconfig.mcux:
```diff
@@ -0,0 +1,14 @@
+
+if HAS_MCUX_FLEXSPI
+
+config MEMC_MCUX_FLEXSPI_HYPERRAM
+	bool "MCUX FlexSPI HyperRAM driver"
+	select MEMC_MCUX_FLEXSPI
```
nit: these lines should be added in commit `drivers: memc: Introduce i.MX RT FlexSPI HyperRAM driver`

pdgendt replied:
Thanks for the review, this got lost in the rebase :)

MaureenHelm commented on drivers/memc/memc_mcux_flexspi.c:
```diff
@@ -6,44 +6,44 @@
 	bool ahb_prefetch;
 	bool ahb_read_addr_opt;
 	bool combination_mode;
+	bool sck_differential_clock;
```
nit: this should be added in commit `drivers: memc: Introduce i.MX RT FlexSPI HyperRAM driver`

MaureenHelm commented on west.yml:
```diff
@@ -100,7 +100,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: b916bca1d5976b0157be80b326e3bb46ef605286
+      revision: pull/83/head
```
This can be updated now that the hal PR has been merged

MaureenHelm reviewed (changes requested):
Thanks @pdgendt !

galak commented on include/drivers/memc/memc_mcux_flexspi.h:
```diff
@@ -0,0 +1,24 @@
+#include <sys/types.h>
+#include <fsl_flexspi.h>
+
+int memc_flexspi_update_lut(const struct device *dev, uint32_t index,
```
Are these APIs meant for application level use?

MaureenHelm replied:
No

pdgendt replied:
Wouldn't the following make more sense in `drivers/flash/CMakeLists.txt`?
```
zephyr_include_directories_ifdef(
  CONFIG_FLASH_MCUX_FLEXSPI_NOR
  ${ZEPHYR_BASE}/drivers/memc
)
```

galak reviewed (changes requested):
Overall looks good, only issue is with ` include/drivers/memc/memc_mcux_flexspi.h`

I think this should be in `drivers/memc/memc_mcux_flexspi.h` and we should add:

`zephyr_include_directories(${ZEPHYR_BASE}/drivers)` to `soc/arm/nxp_imx/rt/CMakeLists.txt`

pdgendt commented:
> Overall looks good, only issue is with ` include/drivers/memc/memc_mcux_flexspi.h`
> 
> I think this should be in `drivers/memc/memc_mcux_flexspi.h` and we should add:
> 
> `zephyr_include_directories(${ZEPHYR_BASE}/drivers)` to `soc/arm/nxp_imx/rt/CMakeLists.txt`

Wouldn't the following make more sense in `drivers/flash/CMakeLists.txt`?
```
zephyr_include_directories_ifdef(
  CONFIG_FLASH_MCUX_FLEXSPI_NOR
  ${ZEPHYR_BASE}/drivers/memc
)
```

galak commented:
> > Overall looks good, only issue is with ` include/drivers/memc/memc_mcux_flexspi.h`
> > I think this should be in `drivers/memc/memc_mcux_flexspi.h` and we should add:
> > `zephyr_include_directories(${ZEPHYR_BASE}/drivers)` to `soc/arm/nxp_imx/rt/CMakeLists.txt`
> 
> Wouldn't the following make more sense in `drivers/flash/CMakeLists.txt`?
> 
> ```
> zephyr_include_directories_ifdef(
>   CONFIG_FLASH_MCUX_FLEXSPI_NOR
>   ${ZEPHYR_BASE}/drivers/memc
> )
> ```

It might, I guess the question is who is calling those APIs?

pdgendt commented:
> It might, I guess the question is who is calling those APIs?

I think only flash or memc drivers will. @MaureenHelm, can you confirm?

MaureenHelm commented:
> > It might, I guess the question is who is calling those APIs?
> 
> I think only flash or memc drivers will. @MaureenHelm, can you confirm?

That's correct
